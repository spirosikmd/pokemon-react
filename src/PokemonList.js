import React, {PureComponent} from 'react';

class PokemonList extends PureComponent {
  handlePokemonClick(url) {
    this.props.onPokemonSelect(url);
  }

  render() {
    return (
      <ul>
        {this.props.pokemons.map(pokemon =>
          <li key={pokemon.name} onClick={() => this.handlePokemonClick(pokemon.url)}>{pokemon.name}</li>
        )}
      </ul>
    );
  }
}

export default PokemonList;
