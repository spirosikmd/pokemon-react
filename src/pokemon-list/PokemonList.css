.ui.pagination.menu {
    padding: 0;
}

.ui.pagination.menu .item:not(.disabled) {
    cursor: pointer;
}

.ui.pagination.menu .item.disabled a {
    color: rgba(40, 40, 40, .3);
}
