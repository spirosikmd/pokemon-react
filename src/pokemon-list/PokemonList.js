import React, { PureComponent } from 'react';
import PokemonRow from '../pokemon-row/PokemonRow';
import ReactPaginate from 'react-paginate';

class PokemonList extends PureComponent {
  constructor(props) {
    super(props);

    this.handlePokemonClick = this.handlePokemonClick.bind(this);
    this.handleFavoriteClick = this.handleFavoriteClick.bind(this);
    this.handlePerPageChange = this.handlePerPageChange.bind(this);
  }

  handlePokemonClick(pokemonName) {
    this.props.onPokemonClick(pokemonName);
  }

  handleFavoriteClick(pokemonName) {
    this.props.onFavoriteClick(pokemonName);
  }

  handlePerPageChange(event) {
    this.props.onPerPageChange(parseInt(event.target.value, 10));
  }

  isFavorite(favoritePokemons, pokemonNameToFind) {
    return (
      favoritePokemons.findIndex(
        favoritePokemon => favoritePokemon.name === pokemonNameToFind
      ) !== -1
    );
  }

  render() {
    const pokemons = this.props.pokemons;
    const favoritePokemons = this.props.favoritePokemons;

    return (
      <div>
        {pokemons.length > 0 &&
          <div>
            <ul>
              {pokemons.map(pokemon => (
                <PokemonRow
                  key={pokemon.name}
                  pokemon={pokemon}
                  isFavorite={this.isFavorite(favoritePokemons, pokemon.name)}
                  onPokemonClick={this.handlePokemonClick}
                  onFavoriteClick={this.handleFavoriteClick}
                />
              ))}
            </ul>

            <label>
              Per page:
              <select
                value={this.props.perPage}
                onChange={this.handlePerPageChange}
              >
                <option value="20">20</option>
                <option value="40">40</option>
                <option value="80">80</option>
              </select>
            </label>

            <ReactPaginate
              pageCount={this.props.pageCount}
              onPageChange={this.props.onPageChange}
              marginPagesDisplayed={1}
              pageRangeDisplayed={3}
            />
          </div>}

        {pokemons.length === 0 &&
          <div>
            <span>No results</span>
            <button onClick={this.props.onFiltersReset}>reset</button>
          </div>}
      </div>
    );
  }
}

export default PokemonList;
