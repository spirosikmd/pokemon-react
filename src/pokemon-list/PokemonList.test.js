import React from 'react';
import renderer from 'react-test-renderer';
import PokemonList from './PokemonList';

const pokemons = [
  {
    name: 'pikachu',
  },
];

const favoritePokemons = [
  {
    name: 'pikachu',
  },
];

const mockHandlePerPageChange = jest.fn();

test('PokemonList renders a list of pokemon and pagination', () => {
  const component = renderer.create(
    <PokemonList
      pokemons={pokemons}
      favoritePokemons={favoritePokemons}
      pageCount={1}
      onPerPageChange={mockHandlePerPageChange}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const select = tree.children[0].children[1].children[1];
  select.props.onChange({
    target: {
      value: '20',
    },
  });
  expect(mockHandlePerPageChange.mock.calls.length).toBe(1);
  expect(mockHandlePerPageChange.mock.calls[0][0]).toBe(20);
});

test('PokemonList isFavorite returns true when pokemon name is part of favorite', () => {
  const pokemonList = new PokemonList();
  expect(pokemonList.isFavorite(favoritePokemons, pokemons[0].name)).toBe(true);
});

test('PokemonList isFavorite returns false when pokemon name is not part of favorite', () => {
  const pokemonList = new PokemonList();
  expect(pokemonList.isFavorite(favoritePokemons, 'bulbasar')).toBe(false);
});
