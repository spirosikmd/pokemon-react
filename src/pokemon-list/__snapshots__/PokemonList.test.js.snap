exports[`test PokemonList renders a list of pokemon and pagination 1`] = `
<div>
  <div>
    <ul>
      <li>
        <span
          onClick={[Function]}
          style={
            Object {
              "color": "red",
            }
          }>
          pikachu
        </span>
        <button
          onClick={[Function]}>
          -
        </button>
      </li>
    </ul>
    <label>
      Per page:
      <select
        onChange={[Function]}
        value={undefined}>
        <option
          value="20">
          20
        </option>
        <option
          value="40">
          40
        </option>
        <option
          value="80">
          80
        </option>
      </select>
    </label>
    <ul
      className={undefined}>
      <li
        className="previous disabled">
        <a
          className={undefined}
          href={undefined}
          onClick={[Function]}
          onKeyPress={[Function]}
          tabIndex="0">
          Previous
        </a>
      </li>
      <li
        className="selected">
        <a
          aria-current="page"
          aria-label="Page 1 is your current page"
          className={undefined}
          href={undefined}
          onClick={[Function]}
          onKeyPress={[Function]}
          tabIndex="0">
          1
        </a>
      </li>
      <li
        className="next disabled">
        <a
          className={undefined}
          href={undefined}
          onClick={[Function]}
          onKeyPress={[Function]}
          tabIndex="0">
          Next
        </a>
      </li>
    </ul>
  </div>
</div>
`;
